### PR TITLE
Set env_config value and add Serialize/Deserialize for Context

### LINKED ISSUE #260: Serialize/deserialize Context
I'd like enable serialization / deserialization of Context to enable passing off handler's input/output for remote execution over the network.

I use a proxy handler that sends the request to AWS SQS queue and then reads a response from another queue. It acts as a proxy for the actual lambda running on my laptop. It's a very handy way of debugging it locally while being fully plugged into the AWS workflow. A more detailed explanation is in https://github.com/rimutaka/lambda-debug-proxy and https://dev.to/rimutaka/debugging-lambdas-on-aws-from-your-local-computer-4j59.

**The changes are superficial**: 
- added Config to Context (gotta be done anyway)
- added Serialize, Deserialize to Config and Context
- replaced ? with expect() with detailed error messages in the affected lines of code

**Full diff**: https://github.com/awslabs/aws-lambda-rust-runtime/compare/master...rimutaka:proxy-experiment

**Branch with the changes**: https://github.com/rimutaka/aws-lambda-rust-runtime/tree/proxy-experiment

This change is unlikely to benefit anyone outside of this specific use case, so it would be logical to make it a feature, but the amount of code added by the compiler is minimal, sometime negative. I tested this by compiling our examples for release.

 Example | proxy | master | diff 
----------|----|------|--------
 basic | 6506808 | 6520728 | -13,920 
 error-handling | 6612224 | 6633936 | -21,712 
 hello-http | 6736936 | 6730992 | 5,944 
 hello-http-without-macros | 6755200 | 6748960 | 6,240 
 macro | 6450784 | 6460960 | -10,176 

I am at a loss why some builds from proxy branch are smaller than master when they have Ser/Deser in them. What I know is that I've been using this proxy branch for weeks with no problems. What did I miss?

The proxy branch is ahead of master and is based on PR #244, so it cannot be merged. I am maintaining it in my own fork and will open a new PR once #244 is resolved. This issue is just for the record.

### PR DESCRIPTION
**Pull request by rimutaka**
Closes #260 

*Description of changes:*

1. Initialize the value for `Context::env_config` which was declared in master many commits ago, but was never set to anything
2. Add `#[derive(Serialize, Deserialize)]` to `Context` and `Config` structs
3. Mock env vars for `successful_end_to_end_run` test function
4. Add more verbose error logging for `Config::from_env`
5. Bumped the version to `0.3.1`

This change enables proxy-ing lambda requests via transports like AWS SQS for local debugging.
Adding Ser/Deser does not affect the binary size unless those features actually used, so there is no point making it an optional feature.

**Backward compatibility**: full

See issue #260 for background discussion and https://github.com/rimutaka/lambda-debug-proxy for a proxy example using this feature.

By submitting this pull request

- [x] I confirm that my contribution is made under the terms of the Apache 2.0 license.
- [x] I confirm that I've made a best effort attempt to update all relevant documentation.

### CODE REVIEW

**Comment by dspatoulas** [comment]
We are having a similar issue using the v0.3.0 version of the runtime in a local or non AWS lambda environment. I can confirm without these changes (or setting the environment variables in docker-compose), when running the `aws-lambda-rust-runtime` in the official AWS Lambda container image (`public.ecr.aws/lambda/provided:al2`) using the examples in the repo causes the runtime to panic. 

We planned to open a similar merge request in the near future, but will hold off until the maintainers have a chance to respond to this MR.

File: README.md [hunk 1]
```diff
@@ -155,6 +155,10 @@ $ unzip -o \
 # Ctrl-D to yield control back to your function
 ```
 
+### Debugging
+
+Lambdas can be run and debugged locally using a special [Lambda debug proxy](https://github.com/rimutaka/lambda-debug-proxy), which is a Lambda function that forwards incoming requests to one AWS SQS queue and reads responses from another queue. A local proxy running on your development computer reads the queue, calls your lambda locally and sends back the response. This approach allows debugging of Lambda functions locally while being part of your AWS workflow. The lambda handler code does not need to be modified between the local and AWS versions.
```
Comment (bahildebrand) [inline README.md:160]: This looks like an awesome project! Would you mind mentioning that it's not an Amazon managed project somewhere in this paragraph?

File: lambda-runtime/Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "lambda_runtime"
-version = "0.3.0"
+version = "0.3.1"
```
Comment (bahildebrand) [inline lambda-runtime/Cargo.toml:3]: Can we take the version bump out of this PR? We usually do it as a part of the release process.

**Review by bahildebrand — CHANGES_REQUESTED** [review]
Overall this looks good. I think we might want to add some automated tests to prevent a regression like this in the future. I'll open an issue and look into it.

File: README.md (hunk 1 above)
Comment (rimutaka) [reply README.md]: See https://github.com/awslabs/aws-lambda-rust-runtime/pull/308/commits/b7a2be806c0acbbe9e239624ac6e6dae8dc6177b.

File: lambda-runtime/Cargo.toml (hunk 2 above)
Comment (rimutaka) [reply lambda-runtime/Cargo.toml]: Done. See https://github.com/awslabs/aws-lambda-rust-runtime/pull/308/commits/317a362f243f60e4670bc3704135c3bfd1f20131

**Review by bahildebrand — APPROVED** [review]
LGTM

**Pull request merged**